crates/artifact_content/src/file_system_state.rs

```diff
@@ -0,0 +1,530 @@
+                .root_files
+                .get(file_name)
+                .map(|(_old_content, old_hash)| old_hash != new_hash)
+                .unwrap_or(true);
+
+            if should_write {
+                operations.push(FileSystemOperation::WriteFile(
+                    file_path,
+                    new_content.clone(),
+                ));
+            }
+        }
+
+        for (server_object, selectables) in &self.nested_files {
+            let server_object_path = artifact_directory.join(server_object.to_string());
+
+            for (selectable, files) in selectables {
+                let selectable_path = server_object_path.join(selectable.to_string());
+
+                for file_name in files.keys() {
+                    let exist_in_new = new
+                        .nested_files
+                        .get(server_object)
+                        .and_then(|s| s.get(selectable))
+                        .and_then(|f| f.get(file_name))
+                        .is_some();
+
+                    if !exist_in_new {
+                        let file_path = selectable_path.join(file_name.to_string());
+                        operations.push(FileSystemOperation::DeleteFile(file_path));
+                    }
+                }
+
+                if !new_selectables.contains(&(server_object, selectable)) {
+                    operations.push(FileSystemOperation::DeleteDirectory(selectable_path));
+                }
+            }
+
+            if !new_server_objects.contains(server_object) {
+                operations.push(FileSystemOperation::DeleteDirectory(server_object_path));
+            }
+        }
+
+        for file_name in self.root_files.keys() {
+            if !new.root_files.contains_key(file_name) {
+                let file_path = artifact_directory.join(file_name.to_string());
+                operations.push(FileSystemOperation::DeleteFile(file_path));
+            }
+        }
+
+        return operations;
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use common_lang_types::{
+        ArtifactPath, ParentObjectEntityNameAndSelectableName, SelectableName,
+        ServerObjectEntityName,
+    };
+    use intern::string_key::Intern;
+    use std::path::PathBuf;
+
+    fn create_artifact(
+        server: Option<&str>,
+        selectable: Option<&str>,
+        file_name: &str,
+        content: &str,
+    ) -> ArtifactPathAndContent {
+        let type_and_field = match (server, selectable) {
+            (Some(s), Some(sel)) => Some(ParentObjectEntityNameAndSelectableName {
+                parent_object_entity_name: ServerObjectEntityName::from(s.intern()),
+                selectable_name: SelectableName::from(sel.intern()),
+            }),
+            _ => None,
+        };
+
+        ArtifactPathAndContent {
+            artifact_path: ArtifactPath {
+                type_and_field,
+                file_name: ArtifactFileName::from(file_name.intern()),
```

Comment:
`file_name.intern().into()`

Reply:
Hmm the compiler complains: `type must be known at this point`